### PR TITLE
LBAC-39 : MIgration of Patients to a Location

### PR DESCRIPTION
**Pull request by VANKINEENITAWRUN**
## Description of what I changed

Added an End-point with /patient-migration a post request with patient's list and location as 
`{
    "locationUuid": "2131aff8-2e2a-480a-b7ab-4ac53250262b",  
    "patientList": ["c5da9efc-3470-4591-bf20-559e5e79f7e7","661e8829-aae1-4e48-b389-819205acef0b","5ab211f5-f461-4239-b2dc-43a83ea3e595"]
}` 
will migrate patients to that Location.

## Issue I worked on

see https://issues.openmrs.org/browse/LBAC-39

## Checklist: I completed these to help reviewers :)

- [x] My pull request only contains **ONE single commit**
(the number above, next to the 'Commits' tab is 1).

  No? -> [read here](https://wiki.openmrs.org/display/docs/Pull+Request+Tips) on how to squash multiple commits into one

- [x] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [x] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [x] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [x] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [x] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

**Comment by VANKINEENITAWRUN** [comment]
@suthagar23 can you please review this PR and merge else working on other Endpoints will create a merge conflict as changes or common in pom files.

**Comment by suthagar23** [comment]
@VANKINEENITAWRUN Let me do it today night

File: omod/src/main/java/org/openmrs/module/locationbasedaccess/web/controller/PatientsMigrationController.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+import org.springframework.web.bind.annotation.RequestMethod;
+import org.springframework.web.bind.annotation.ResponseBody;
+import org.springframework.web.bind.annotation.ResponseStatus;
+
+@Controller
+@RequestMapping(value = "/rest/" + RestConstants.VERSION_1 + "/patient-migration")
+public class PatientsMigrationController extends BaseRestController {
+
+	@Qualifier("patientService")
+	@Autowired
```
Comment (suthagar23) [inline omod/src/main/java/org/openmrs/module/locationbasedaccess/web/controller/PatientsMigrationController.java]: Field autowiring isn't suggested. Can you do it through constructor autowiring?

File: omod/src/main/java/org/openmrs/module/locationbasedaccess/web/controller/PatientsMigrationController.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+	private PatientService patientService;
+
+	@RequestMapping(method = RequestMethod.POST)
+	@ResponseBody
+	@ResponseStatus(value = HttpStatus.OK)
+	public void migratePatientLocation(@RequestBody Map<String, Object> body){
+		List<String> patientList = (List<String>) body.get("patientList");
+		String locationUuid = (String)body.get("locationUuid");
```
Comment (suthagar23) [inline omod/src/main/java/org/openmrs/module/locationbasedaccess/web/controller/PatientsMigrationController.java]: What will happen if this locationUuid is invalid?

File: omod/src/main/java/org/openmrs/module/locationbasedaccess/web/controller/PatientsMigrationController.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+import org.springframework.stereotype.Controller;
+import org.springframework.web.bind.annotation.ExceptionHandler;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestMethod;
+import org.springframework.web.bind.annotation.ResponseBody;
+import org.springframework.web.bind.annotation.ResponseStatus;
+
+@Controller
+@RequestMapping(value = "/rest/" + RestConstants.VERSION_1 + "/patient-migration")
```
Comment (suthagar23) [inline omod/src/main/java/org/openmrs/module/locationbasedaccess/web/controller/PatientsMigrationController.java]: Haven't you added swagger docs for this API?

**Comment by suthagar23** [comment]
What do you think about this api design?
`<url>/lbac/patients-location`

File: omod/src/main/java/org/openmrs/module/locationbasedaccess/web/controller/PatientsMigrationController.java [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+import org.springframework.http.HttpStatus;
+import org.springframework.stereotype.Controller;
+import org.springframework.web.bind.annotation.ExceptionHandler;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestMethod;
+import org.springframework.web.bind.annotation.ResponseBody;
+import org.springframework.web.bind.annotation.ResponseStatus;
+
+@Controller
```
Comment (suthagar23) [inline omod/src/main/java/org/openmrs/module/locationbasedaccess/web/controller/PatientsMigrationController.java]: Add java-docs

File: omod/src/main/java/org/openmrs/module/locationbasedaccess/web/controller/PatientsMigrationController.java [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+	@RequestMapping(method = RequestMethod.POST)
+	@ResponseBody
+	@ResponseStatus(value = HttpStatus.OK)
+	public void migratePatientLocation(@RequestBody Map<String, Object> body){
+		List<String> patientList = (List<String>) body.get("patientList");
+		String locationUuid = (String)body.get("locationUuid");
+		String locationAttributeUuid = Context.getAdministrationService().getGlobalProperty(
+				LocationBasedAccessConstants.LOCATION_ATTRIBUTE_GLOBAL_PROPERTY_NAME);
+		if (StringUtils.isNotBlank(locationAttributeUuid)) {
```
Comment (suthagar23) [inline omod/src/main/java/org/openmrs/module/locationbasedaccess/web/controller/PatientsMigrationController.java]: If it's blank, what kind of response will you send?

File: omod/src/main/java/org/openmrs/module/locationbasedaccess/web/controller/PatientsMigrationController.java [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+			final PersonAttributeType personAttributeType = Context.getPersonService().getPersonAttributeTypeByUuid(locationAttributeUuid);
+			PersonAttribute personAttribute = new PersonAttribute(personAttributeType, locationUuid);
+			for(Iterator<String> iterator = patientList.iterator(); iterator.hasNext();){
```
Comment (suthagar23) [inline omod/src/main/java/org/openmrs/module/locationbasedaccess/web/controller/PatientsMigrationController.java]: Format this line

File: omod/src/main/java/org/openmrs/module/locationbasedaccess/web/controller/PatientsMigrationController.java [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+public class PatientsMigrationController extends BaseRestController {
+
+	@Qualifier("patientService")
+	@Autowired
+	private PatientService patientService;
+
+	@RequestMapping(method = RequestMethod.POST)
+	@ResponseBody
+	@ResponseStatus(value = HttpStatus.OK)
+	public void migratePatientLocation(@RequestBody Map<String, Object> body){
```
Comment (suthagar23) [inline omod/src/main/java/org/openmrs/module/locationbasedaccess/web/controller/PatientsMigrationController.java]: format this line

File: omod/src/main/java/org/openmrs/module/locationbasedaccess/web/controller/PatientsMigrationController.java [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+		List<String> patientList = (List<String>) body.get("patientList");
```
Comment (suthagar23) [inline omod/src/main/java/org/openmrs/module/locationbasedaccess/web/controller/PatientsMigrationController.java]: What happens if patients list isn't an array? - Make sure about the exceptions if possible

File: omod/src/main/java/org/openmrs/module/locationbasedaccess/web/controller/PatientsMigrationController.java (hunk 3 above)
Comment (VANKINEENITAWRUN) [reply omod/src/main/java/org/openmrs/module/locationbasedaccess/web/controller/PatientsMigrationController.java]: @suthagar23  we are not using any model in this so we are not able to use the OpenMRS swagger, for this we need to add swagger separately for our module. doing this is good once we are done with all the rest endpoints else adding swagger configuration in every PR individually will make a mess up I feel. Please provide your insights.

**Comment by dkayiwa** [comment]
@suthagar23 any more comments for this?

**Comment by suthagar23** [comment]
@dkayiwa Still some comments need to be addressed here.

@VANKINEENITAWRUN Can you go through all the comments, and reply or fix?

File: omod/src/main/java/org/openmrs/module/locationbasedaccess/web/controller/PatientsMigrationController.java (hunk 4 above)
Comment (VANKINEENITAWRUN) [reply omod/src/main/java/org/openmrs/module/locationbasedaccess/web/controller/PatientsMigrationController.java]: Added

**Comment by dkayiwa** [comment]
@VANKINEENITAWRUN did you address the above?

File: omod/src/main/java/org/openmrs/module/locationbasedaccess/web/controller/PatientsMigrationController.java (hunk 8 above)
Comment (suthagar23) [reply omod/src/main/java/org/openmrs/module/locationbasedaccess/web/controller/PatientsMigrationController.java]: @VANKINEENITAWRUN

File: omod/src/test/java/org/openmrs/module/locationbasedaccess/web/controller/PatientsMigrationController_Test.java [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+	private final int PATIENT2_ID =2;
+	private final int PATIENT3_ID =3;
+
+	@Before
+	public void init() throws Exception {
+		patientService = Context.getPatientService();
+		executeDataSet(XML_FILENAME_WITH_PERSON_DATA);
+	}
+
+	private String getURI(){
```
Comment (suthagar23) [inline omod/src/test/java/org/openmrs/module/locationbasedaccess/web/controller/PatientsMigrationController_Test.java]: fix the code style issue

File: omod/src/main/java/org/openmrs/module/locationbasedaccess/web/controller/PatientsMigrationController.java (hunk 2 above)
Comment (VANKINEENITAWRUN) [reply omod/src/main/java/org/openmrs/module/locationbasedaccess/web/controller/PatientsMigrationController.java]: location_notValid

@suthagar23  Fixed it.

File: omod/src/main/java/org/openmrs/module/locationbasedaccess/web/controller/PatientsMigrationController.java (hunk 8 above)
Comment (VANKINEENITAWRUN) [reply omod/src/main/java/org/openmrs/module/locationbasedaccess/web/controller/PatientsMigrationController.java]: patientListException
@suthagar23  Fixed.

File: omod/src/main/java/org/openmrs/module/locationbasedaccess/web/controller/PatientsMigrationController.java (hunk 5 above)
Comment (VANKINEENITAWRUN) [reply omod/src/main/java/org/openmrs/module/locationbasedaccess/web/controller/PatientsMigrationController.java]: Fixed it. Can review Now.

**Pull request merged**